### PR TITLE
Simplify keep alive logic

### LINKED ISSUE #21067: Simplify service worker keep-alive logic
## Context

For our purposes at MetaMask, one of the most significant differences between manifest v3 and v2 extension implementations was that the background process runs on a service worker that restarts every 5 minutes. However, [we recently learned](https://consensys.slack.com/archives/C02DUDT2WTU/p1693402457225629) that Chrome dropped that requirement, which means the mv3-specific code still needs to persist in the service worker state, but the implementation can be simplified from what we have previously implemented.

## Approach

Simplify code on `app/scripts/contentscript.js` `background.js` and `ui.js`, namely removing keep alive and acknowledgment code that is no longer needed. The official recommendation for how to keep the extension process alive indefinitely and to save a timestamp to `chrome.storage` every X seconds. We can use `chrome.storage.session` has [it can store up to 10 Mb](https://developer.chrome.com/docs/extensions/reference/storage/#property-session).

### `contentscript.js`

Delete the following variables and methods:
  - `EXTENSION_CONTEXT_INVALIDATED_CHROMIUM_ERROR`,
  - `WORKER_KEEP_ALIVE_INTERVAL`, 
  - `WORKER_KEEP_ALIVE_MESSAGE`,
  - `TIME_45_MIN_IN_MS`,
  - `IGNORE_INIT_METHODS_FOR_KEEP_ALIVE`,
  - `keepAliveInterval`,
  - `keepAliveTimer`,
  - `sendMessageWorkerKeepAlive`,
  - `runWorkerKeepAliveInterval`

### `ui.js`

- `lastMessageReceivedTimestamp`
- `ackTimeoutToDisplayError`
- `ackKeepAliveListener`
- `keepAliveInterval`
- `WORKER_KEEP_ALIVE_INTERVAL`
- `WORKER_KEEP_ALIVE_MESSAGE`
- `ACK_KEEP_ALIVE_WAIT_TIME`
- `ACK_KEEP_ALIVE_MESSAGE`

We can also remove two of the ignoreStreams statements on `app/scripts/lib/stream-utils.js`.

### `background.js`

Remove the logic that listens for the `WORKER_KEEP_ALIVE_MESSAGE` and returns an `ACK_KEEP_ALIVE_MESSAGE`.

Add the following method

```js
function saveTimestamp() {
  const timestamp = new Date().toISOString();

  browser.storage.session.set({ timestamp }, () => {
    console.log('timestamp saved to `browser.storage.session`', timestamp);
  });
}
```

And call the method repeatedly inside `initialize()`.
```js
// Save the timestamp immediately and then every `SAVE_TIMESTAMP_INTERVAL`
// miliseconds. This keeps the service worker alive.
const SAVE_TIMESTAMP_INTERVAL_MS = 30 * 1000;

saveTimestamp();
setInterval(saveTimestamp, SAVE_TIMESTAMP_INTERVAL_MS);
```

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBLzhzQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--70de7b4060a706fd8b90cd6be310afcdefc2744d/Screenshot%202023-10-12%20at%2013.44.58.png)

### Deprecate `service_worker_restarted_time`

See [PR](https://github.com/MetaMask/metamask-extension/pull/18346)

## Acceptance Criteria
- Create a new function that saves a timestamp to localStorage every `30` seconds.
- Call that function in `initialize` inside background.js with the appropriate check for the mv3 environment variable.
- Remove methods and variables described above
- Fix the console error described above

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## Description

For our purposes at MetaMask, one of the most significant differences between manifest v3 and v2 extension implementations was that the background process runs on a service worker that restarts every 5 minutes. However, [we recently learned](https://consensys.slack.com/archives/C02DUDT2WTU/p1693402457225629) that Chrome dropped that requirement, which means the mv3-specific code still needs to persist in the service worker state, but the implementation can be simplified from what we have previously done.

The approach has been to simplify code on `app/scripts/contentscript.js` `background.js` and `ui.js`, namely removing keep alive and acknowledgment code that is no longer needed.

## **Manual testing steps**

1. Build the extension through the manifest v3 file using `yarn dist:mv3` or `yarn start:mv3`
2. Load the extension and perform a few user actions such as adding a new account.
3. Wait 5 minutes (the previous limit for service worker lifecycle.
4. Perform a few user actions such as adding a new account
5. The actions should be performed and the app should not crash

## **Related issues**

Fixes #21067

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/background.js [hunk 1]
```diff
@@ -208,6 +205,14 @@ browser.runtime.onConnectExternal.addListener(async (...args) => {
   connectExternal(...args);
 });
 
+function saveTimestamp() {
+  const timestamp = new Date().toISOString();
+
+  browser.storage.session.set({ timestamp }, () => {
+    console.log('Timestamp saved to `browser.storage.session`', timestamp);
```
Comment (brad-decker) [inline app/scripts/background.js:212]: Do we need this console.log for functionality or should it be removed?

File: app/scripts/background.js [hunk 2]
```diff
@@ -928,6 +927,8 @@ function setupSentryGetStateGlobal(store) {
 
 function initBackground() {
   initialize().catch(log.error);
+
+  console.log('after initialize');
```
Comment (brad-decker) [inline app/scripts/background.js:931]: Safe to remove?

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
     "start:mmi": "yarn start --build-type mmi",
     "start:lavamoat": "yarn build:dev dev --apply-lavamoat=true",
     "dist": "yarn build dist",
+    "dist:mv3": "ENABLE_MV3=true yarn build dist",
```
Comment (DDDDDanica) [inline package.json:16]: Maybe we should move this to a separate MR to make sure the functionality is related?

File: app/scripts/background.js [hunk 4]
```diff
@@ -208,6 +205,14 @@ browser.runtime.onConnectExternal.addListener(async (...args) => {
   connectExternal(...args);
 });
 
+function saveTimestamp() {
+  const timestamp = new Date().toISOString();
+
+  browser.storage.session.set({ timestamp }, () => {
```
Comment (DDDDDanica) [inline app/scripts/background.js:211]: Quick question: could be me overthinking, but what if we clear the session storage, and i think the app will still be running right?

File: app/scripts/background.js [hunk 5]
```diff
@@ -300,6 +305,13 @@ async function initialize() {
 
     let isFirstMetaMaskControllerSetup;
     if (isManifestV3) {
+      // Save the timestamp immediately and then every `SAVE_TIMESTAMP_INTERVAL`
+      // miliseconds. This keeps the service worker alive.
+      const SAVE_TIMESTAMP_INTERVAL_MS = 30 * 1000;
```
Comment (pedronfigueiredo) [inline app/scripts/background.js:310]: Increase the frequency at which the timestamp is saved

File: package.json [hunk 6]
```diff
@@ -263,7 +264,7 @@
     "@metamask/message-manager": "^7.3.0",
     "@metamask/metamask-eth-abis": "^3.0.0",
     "@metamask/name-controller": "^3.0.0",
-    "@metamask/network-controller": "^12.2.0",
+    "@metamask/network-controller": "^15.0.0",
```
Comment (jiexi) [inline package.json:267]: This is blocked by v13 bump here https://github.com/MetaMask/metamask-extension/pull/20652

Comment (pedronfigueiredo) [reply package.json]: This was inadvertently committed on this PR, removing it. thanks for flagging it though

**Review by danjm — APPROVED** [review]
LGTM

**Review by DDDDDanica — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
